### PR TITLE
TransportNodesListGatewayStartedShards should fall back to disk based index metadata if not found in cluster state

### PR DESCRIPTION
**Pull request by bleskes**
When an index is recovered from disk it's metadata is imported first and the master reaches out to the nodes looking for shards of that index. Sometimes those requests reach other nodes before the cluster state is processed by them. At the moment, that situation disables the checking of the store, which requires the meta data (indices with custom path need to know where the data is). When corruption hits this means we may assign a shard to node with corrupted store, which will be caught later on but causes confusion. Instead we can try loading the meta data from disk in those cases.

Relates to #17630

### CODE REVIEW

File: core/src/test/java/org/elasticsearch/gateway/RecoveryFromGatewayIT.java [hunk 1]
```diff
@@ -72,20 +78,20 @@ public void testOneNodeRecoverFromGateway() throws Exception {
         internalCluster().startNode();
 
         String mapping = XContentFactory.jsonBuilder().startObject().startObject("type1")
-                .startObject("properties").startObject("appAccountIds").field("type", "text").endObject().endObject()
-                .endObject().endObject().string();
+            .startObject("properties").startObject("appAccountIds").field("type", "text").endObject().endObject()
```
Comment (dakrone) [inline core/src/test/java/org/elasticsearch/gateway/RecoveryFromGatewayIT.java]: So many extra changes to this file :(

File: core/src/main/java/org/elasticsearch/gateway/TransportNodesListGatewayStartedShards.java [hunk 2]
```diff
@@ -110,47 +120,58 @@ protected NodesGatewayStartedShards newResponse(Request request, AtomicReference
...
+                IndexMetaData metaData = clusterService.state().metaData().index(shardId.getIndex());
+                if (metaData == null) {
+                    // we may send this requests while processing the cluster state that recovered the index
+                    // sometimes the request comes in before the local node processed that cluster state
+                    // in such cases we can load it from disk
+                    metaData = IndexMetaData.FORMAT.loadLatestState(logger, nodeEnv.indexPaths(shardId.getIndex()));
                 }
-                // old shard metadata doesn't have the actual index UUID so we need to check if the actual uuid in the metadata
-                // is equal to IndexMetaData.INDEX_UUID_NA_VALUE otherwise this shard doesn't belong to the requested index.
-                if (indexUUID.equals(shardStateMetaData.indexUUID) == false
```
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/gateway/TransportNodesListGatewayStartedShards.java]: Does this happen post-rename then? So it's okay to remove these uuid checks?

Comment (bleskes) [inline core/src/main/java/org/elasticsearch/gateway/TransportNodesListGatewayStartedShards.java]: yes, the folders are renamed before we open the transport gates. We now verify the uuid in the cluster state metadata and/or read the meta data file from uuid'ed folder which I think make this test redundant. On top of it, if we get it wrong, the allocation id will not match what the master is looking for.

**Comment by dakrone** [comment]
LGTM, thanks Boaz

**Pull request merged**